lib/SPIRV/LLVMBitreverse.h

```diff
@@ -59,8 +62,70 @@
 //     ...
 //     ret <4 x i8> %or12
 //   }
+#define GEN_CONST1(BASE_TYPE, VAL) #VAL
+#define GEN_CONST2(BASE_TYPE, VAL)                                             \
+  "<" #BASE_TYPE " " #VAL "," #BASE_TYPE " " #VAL ">"
+#define GEN_CONST3(BASE_TYPE, VAL)                                             \
+  "<" #BASE_TYPE " " #VAL "," #BASE_TYPE " " #VAL "," #BASE_TYPE " " #VAL ">"
+#define GEN_CONST4(BASE_TYPE, VAL)                                             \
+  "<" #BASE_TYPE " " #VAL "," #BASE_TYPE " " #VAL "," #BASE_TYPE " " #VAL      \
+  "," #BASE_TYPE " " #VAL ">"
+#define GEN_CONST8(BASE_TYPE, VAL)                                             \
+  "<" #BASE_TYPE " " #VAL "," #BASE_TYPE " " #VAL "," #BASE_TYPE " " #VAL      \
+  "," #BASE_TYPE " " #VAL "," #BASE_TYPE " " #VAL "," #BASE_TYPE " " #VAL      \
+  "," #BASE_TYPE " " #VAL "," #BASE_TYPE " " #VAL ">"
+#define GEN_CONST16(BASE_TYPE, VAL)                                            \
+  "<" #BASE_TYPE " " #VAL "," #BASE_TYPE " " #VAL "," #BASE_TYPE " " #VAL      \
+  "," #BASE_TYPE " " #VAL "," #BASE_TYPE " " #VAL "," #BASE_TYPE " " #VAL      \
+  "," #BASE_TYPE " " #VAL "," #BASE_TYPE " " #VAL "," #BASE_TYPE " " #VAL      \
+  "," #BASE_TYPE " " #VAL "," #BASE_TYPE " " #VAL "," #BASE_TYPE " " #VAL      \
+  "," #BASE_TYPE " " #VAL "," #BASE_TYPE " " #VAL "," #BASE_TYPE " " #VAL      \
+  "," #BASE_TYPE " " #VAL ">"
+
+// clang-format off
+#define MAKE_BITREVERSE_2BIT(SUFFIX,TYPE_STR,NUM_ELTS,BASE_TYPE)                     \
+static const char LLVMBitreverse ## SUFFIX[]{"                                     \n\
+define " TYPE_STR " @llvm_bitreverse_" #SUFFIX "(" TYPE_STR " %A) {                \n\
+entry:                                                                             \n\
+  %and  = shl         " TYPE_STR " %A, " GEN_CONST ## NUM_ELTS(BASE_TYPE,1) "      \n\
+  %shr4 = lshr        " TYPE_STR " %A, " GEN_CONST ## NUM_ELTS(BASE_TYPE,1) "      \n\
+  %or   = or disjoint " TYPE_STR " %and, %shr4                                     \n\
+  ret                 " TYPE_STR " %or                                             \n\
+}                                                                                  \n\
+"}
+
+MAKE_BITREVERSE_2BIT(i2,    "i2",         1, i2);
+MAKE_BITREVERSE_2BIT(v2i2,  "<2 x i2>",   2, i2);
+MAKE_BITREVERSE_2BIT(v3i2,  "<3 x i2>",   3, i2);
+MAKE_BITREVERSE_2BIT(v4i2,  "<4 x i2>",   4, i2);
+MAKE_BITREVERSE_2BIT(v8i2,  "<8 x i2>",   8, i2);
+MAKE_BITREVERSE_2BIT(v16i2, "<16 x i2>", 16, i2);
+
+#define MAKE_BITREVERSE_4BIT(SUFFIX,TYPE_STR,NUM_ELTS,BASE_TYPE)                     \
+static const char LLVMBitreverse ## SUFFIX[]{"                                     \n\
+define " TYPE_STR " @llvm_bitreverse_" #SUFFIX "(" TYPE_STR " %A) {                \n\
+entry:                                                                             \n\
+  %and  = shl         " TYPE_STR " %A,    " GEN_CONST ## NUM_ELTS(BASE_TYPE, 2) "  \n\
```

Comment:
Just curious why some var names are not aligned with inst opcode while most are.

Reply:
I think the var names were just chosen by clang++.  If you look throughout the file the result of a 'shl' is written to a %and* variable.